proj/hps_accel/gateware/hps_cfu.py

```diff
@@ -71,23 +72,25 @@ def connect_verify_register(self, m, set, get):
 
     def connect_macc(self, m, set, input_store, filter_store, macc, get):
         m.d.comb += macc.offset.eq(set.values[Constants.REG_INPUT_OFFSET])
-        inputs = Cat(input_store.data_output[0].payload,
-                     input_store.data_output[1].payload,
-                     input_store.data_output[2].payload,
-                     input_store.data_output[3].payload)
-        for n, val in enumerate(all_words(inputs, 8)):
-            m.d.comb += macc.inputs[n].eq(val)
-        filters = Cat(filter_store.output[0].payload,
-                      filter_store.output[1].payload,
-                      filter_store.output[2].payload,
-                      filter_store.output[3].payload)
-        for n, val in enumerate(all_words(filters, 8)):
-            m.d.comb += macc.filters[n].eq(val)
+
+        # Join all 4 input value streams and all 4 filter value streams
+        # into a single stream holding all operands.
+        # Then connect that to the macc.
+        operands_layout = Layout(
+                [(f'input_{i}', unsigned(32)) for i in range(4)] +
```

Comment:
Do you mean, you would rather a list of the 8 explicit layout fields (and 8 explicit corresponding `connect()` calls below), instead of using range(4) to produce the four numbered fields?

I did actually have that in one version, but it felt like a bit too much repetitive copy-paste.

The idea of a subclass of ConcatenatingBuffer specialised to the specific layout, is interesting but I guess it's a slightly different issue, more related to your point below about the API of its constructor.

Reply:
range(4) is good! I'm more flagging the construction of strings with index values encoded into them. This is an Ok thing to do when there is no alternative, but an different API would avoid the need for it.